src/ast/str_util.rs

```diff
@@ -0,0 +1,306 @@
+            parent.syntax().children_with_tokens().filter_map(SyntaxElement::into_token)
+        }
+
+        let mut parts = Vec::new();
+        let mut literals = 0;
+        let mut common = std::usize::MAX;
+        let multiline = children_tokens(self).next().map_or(false, |t| t.text() == "''");
+        let mut last_was_ast = false;
+
+        for child in self.syntax().children_with_tokens() {
+            match &child {
+                NodeOrToken::Token(token) if token.kind() == TOKEN_STRING_CONTENT => {
+                    let text: &str = token.text();
+
+                    let line_count = text.lines().count();
+                    let next_is_ast = child
+                        .next_sibling_or_token()
+                        .map_or(false, |child| child.kind() == NODE_STRING_INTERPOL);
+                    for (i, line) in text.lines().enumerate().skip(if last_was_ast { 1 } else { 0 })
+                    {
+                        let indent: usize = indention(line).count();
+                        if (i != line_count - 1 || !next_is_ast) && indent == line.chars().count() {
+                            // line is empty and not the start of an
+                            // interpolation, ignore indention
+                            continue;
+                        }
+                        common = common.min(indent);
+                    }
+                    parts.push(StrPart::Literal(text.to_string()));
+                    literals += 1;
+                }
+                NodeOrToken::Token(token) => {
+                    assert!(token.kind() == TOKEN_STRING_START || token.kind() == TOKEN_STRING_END)
+                }
+                NodeOrToken::Node(node) => {
+                    assert_eq!(node.kind(), NODE_STRING_INTERPOL);
+                    parts.push(StrPart::Ast(ast::StrInterpol::cast(node.clone()).unwrap()));
+                    last_was_ast = true;
+                }
+            }
+        }
+
+        let mut i = 0;
+        for part in parts.iter_mut() {
+            if let StrPart::Literal(ref mut text) = part {
+                if multiline {
+                    *text = remove_indent(text, i == 0, common);
+                    if i == literals - 1 {
+                        // Last index
+                        remove_trailing(text);
+                    }
+                }
+                *text = unescape(text, multiline);
+                i += 1;
+            }
+        }
+
+        parts
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum StrPart {
+    Literal(String),
+    Interpolation(ast::StrInterpol),
+}
+
+/// Interpret escape sequences in the nix string and return the converted value
+pub fn unescape(input: &str, multiline: bool) -> String {
+    let mut output = String::new();
+    let mut input = input.chars().peekable();
+    loop {
+        match input.next() {
+            None => break,
```

Comment:
```suggestion
    while let Some(i) = input.next() {
        match i {
```
and remove `Some` from the other branches at this level.